{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "gpuType": "T4",
      "toc_visible": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "GPU"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# Chroma Demo"
      ],
      "metadata": {
        "id": "pTze5gkeWJUH"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "In this demo we will conduct a simplified enzyme enhancement workflow of the IsPETase enzyme. The main goal is to improve the thermal stability of the enzyme while retaining the high PET degradation efficiency.\n",
        "\n",
        "This will be done by retaining the PETase capabilities of the enzyme while de novo generating a backbone; both the docking site and the catalytic triad will be kept fixed. The catalytic triad is composed of Ser160, Asp206, and His237. The docking site is composed of two subsites; subsite I positions one monomer with a benzene ring from PET between Tyr 87 and Trp185 and subsite II is used to hold three more monomers, and exists of amino acids including Thr88, Ala89, Trp159, Ile232, Asn233, Ser236, Ser238, Asn241, Asn244, Ser245, Asn246, and Arg280."
      ],
      "metadata": {
        "id": "IzNGBMvVJK9K"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Installing Chroma"
      ],
      "metadata": {
        "id": "1bBxAlIbJWw9"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "collapsed": true,
        "id": "cblMwIDEWILW"
      },
      "outputs": [],
      "source": [
        "#Install condacolab to battle the dependency issues\n",
        "!pip install -q condacolab\n",
        "import condacolab\n",
        "condacolab.install()"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#Optional cell to check\n",
        "#When this cell fails, run it again, I have no idea why this error occurs but it does, and just by running it again it works\n",
        "import condacolab\n",
        "condacolab.check()"
      ],
      "metadata": {
        "id": "Zw9c47JrgNhQ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#Make the environment\n",
        "!conda create --name chromademo python=3.9"
      ],
      "metadata": {
        "collapsed": true,
        "id": "MvkXGGbtWTg6"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#Shell cell to install chroma\n",
        "%%shell\n",
        "eval \"$(conda shell.bash hook)\"\n",
        "conda activate chromademo\n",
        "pip install git+https://github.com/generatebio/chroma.git\n",
        "python --version"
      ],
      "metadata": {
        "id": "wEum4Nn_WZbu",
        "collapsed": true
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#General imports needed for chroma\n",
        "import sys\n",
        "import os\n",
        "import contextlib\n",
        "import torch\n",
        "from pathlib import Path\n",
        "\n",
        "#Demo of chroma set preferred encoding, so for safety sake\n",
        "import locale\n",
        "locale.getpreferredencoding = lambda: \"UTF-8\"\n",
        "\n",
        "#Google collab can't find the conda environment, so added to path\n",
        "sys.path.append(\"//usr//local//envs//chromademo//lib//python3.9//site-packages\")\n",
        "\n",
        "#Chroma imports as described in demo\n",
        "from chroma import Chroma, Protein, conditioners\n",
        "from chroma.models import graph_classifier, procap\n",
        "from chroma.utility.api import register_key\n",
        "from chroma.utility.chroma import letter_to_point_cloud, plane_split_protein"
      ],
      "metadata": {
        "id": "8lSNMN5SbYGT",
        "collapsed": true
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#Register the API key and select device\n",
        "api_key = \"\"\n",
        "register_key(api_key)\n",
        "\n",
        "#T4 runtime uses nvidia\n",
        "device = \"cuda\""
      ],
      "metadata": {
        "id": "hjgu7GOqentf"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#Instantiate chroma\n",
        "with contextlib.redirect_stdout(None):\n",
        "    chroma = Chroma()"
      ],
      "metadata": {
        "id": "CxJeWzKYOcUm"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Generating alternative IsPETase enzymes"
      ],
      "metadata": {
        "id": "W2NFXbUbZCgv"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "output_dir_alt = Path(\"alt_proteins\")\n",
        "output_dir_alt.mkdir(exist_ok = True)\n",
        "\n",
        "#Select IsPETase PDB ID\n",
        "PDB_ID = \"5XJH\"\n",
        "\n",
        "#Number of proteins to generate\n",
        "NUM_PROTEINS = 100\n",
        "\n",
        "#List to save all protein sequences\n",
        "alternative_proteins = []\n",
        "\n",
        "#Load protein\n",
        "protein = Protein.from_PDBID(PDB_ID, device = device)\n",
        "\n",
        "#Fix the catalytic site\n",
        "selection_string = \"not (resid 160 or resid 206 or resid 237 or resid 87 or resid 185 or resid 88 or resid 89 or resid 159 or resid 232 or resid 233 or resid 236 or resid 238 or resid 241 or resid 244 or resid 245 or resid 246 or resid 280)\"\n",
        "\n",
        "#SubstructureConditioner\n",
        "substructure_conditioner = conditioners.SubstructureConditioner(\n",
        "    protein=protein,\n",
        "    backbone_model=chroma.backbone_network,\n",
        "    selection=selection_string\n",
        ").to(device)\n",
        "\n",
        "#Generate alternative proteins\n",
        "for i in range(NUM_PROTEINS):\n",
        "    print(f\"Generating protein {i + 1}/{NUM_PROTEINS}\")\n",
        "    alt_protein, trajectories = chroma.sample(\n",
        "        protein_init=protein,\n",
        "        conditioner=substructure_conditioner,\n",
        "        langevin_factor=2.0,\n",
        "        langevin_isothermal=True,\n",
        "        inverse_temperature=8.0,\n",
        "        sde_func=\"langevin\",\n",
        "        steps=200,\n",
        "        full_output=True,\n",
        "    )\n",
        "\n",
        "    #Extract amino acid sequence, and add it to the list\n",
        "    sequence = alt_protein.sequence()\n",
        "    name = f\"alt_protein_{i + 1}\"\n",
        "    alternative_proteins.append((name, sequence))\n",
        "\n",
        "    #Save pdb structure to map\n",
        "    alt_protein.to(f\"alt_proteins/generated_protein_{i + 1}.pdb\")\n",
        "    print(f\"\\nAlternative protein {i + 1}\\n : {alt_protein})\")\n",
        "\n",
        "print(f\"All {NUM_PROTEINS} alternative proteins have been generated!\")"
      ],
      "metadata": {
        "id": "D82gxrGQfIla"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Chroma citation\n",
        "@Article{Chroma2023,\n",
        "  author  = {Ingraham, John B. and Baranov, Max and Costello, Zak and Barber, Karl W. and Wang, Wujie and Ismail, Ahmed and Frappier, Vincent and Lord, Dana M. and Ng-Thow-Hing, Christopher and Van Vlack, Erik R. and Tie, Shan and Xue, Vincent and Cowles, Sarah C. and Leung, Alan and Rodrigues, Jo\\~{a}o V. and Morales-Perez, Claudio L. and Ayoub, Alex M. and Green, Robin and Puentes, Katherine and Oplinger, Frank and Panwar, Nishant V. and Obermeyer, Fritz and Root, Adam R. and Beam, Andrew L. and Poelwijk, Frank J. and Grigoryan, Gevorg},\n",
        "  journal = {Nature},\n",
        "  title   = {Illuminating protein space with a programmable generative model},\n",
        "  year    = {2023},\n",
        "  volume  = {},\n",
        "  number  = {},\n",
        "  pages   = {},\n",
        "  doi     = {10.1038/s41586-023-06728-8}\n",
        "}"
      ],
      "metadata": {
        "id": "fYiStVCmt0-l"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Calculating thermal stability\n",
        "In order to test whether one of the generated alternative proteins are more thermal stable we use the [DeepSTABp](https://csb-deepstabp.bio.rptu.de/) predictor.\n",
        "\n",
        "The online GUI of this algorithm takes as input the FASTA-file of all alternative proteins, the growth temperature, and the environment temperature prediction."
      ],
      "metadata": {
        "id": "o8VLMCfKasCT"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Generate a combined fasta file of all generated alternative proteins\n",
        "fasta_file = output_dir_alt / \"alternative_proteins.fasta\"\n",
        "\n",
        "with open(fasta_file, \"w\") as f:\n",
        "    for name, sequence in alternative_proteins:\n",
        "        f.write(f\">{name}\\n{sequence}\\n\")\n",
        "\n",
        "print(f\"All {len(alternative_proteins)} alternative proteins have been saved to {fasta_file}\")"
      ],
      "metadata": {
        "id": "F7s8G3rdYIKL"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "\n",
        "tm = pd.read_csv(\"20251114_105022_DeepStabP.csv\", header=None, names=[\"name\", \"Tm\"])\n",
        "\n",
        "tm.nlargest(10, \"Tm\")"
      ],
      "metadata": {
        "id": "P8vrk3gZ_S3l"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!zip -r /content/alt_proteins.zip /content/alt_proteins"
      ],
      "metadata": {
        "id": "O85Uu33Wq0DD"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### DeepSTABp citation\n",
        "Jung F, Frey K, Zimmer D, MÃ¼hlhaus T. DeepSTABp: A Deep Learning Approach for the Prediction of Thermal Protein Stability. International Journal of Molecular Sciences. 2023; 24(8):7444. https://doi.org/10.3390/ijms24087444"
      ],
      "metadata": {
        "id": "ChgEmu8zJ0U1"
      }
    }
  ]
}